src/GafferScene/FilterQuery.cpp

```diff
@@ -0,0 +1,346 @@
+	}
+
+	if(
+		input == filterPlug() ||
+		input == scenePlug()->existsPlug()
+	)
+	{
+		outputs.push_back( closestAncestorInternalPlug() );
+	}
+
+	if(
+		input == locationPlug() ||
+		input == closestAncestorInternalPlug()
+	)
+	{
+		outputs.push_back( closestAncestorPlug() );
+	}
+}
+
+void FilterQuery::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	if( output == matchPlug() )
+	{
+		ComputeNode::hash( output, context, h );
+		const string location = locationPlug()->getValue();
+		if( !location.empty() )
+		{
+			ScenePlug::PathScope scope( context, ScenePlug::stringToPath( location ) );
+			if( scenePlug()->existsPlug()->getValue() )
+			{
+				filterPlug()->hash( h );
+			}
+		}
+	}
+	else if( output == exactMatchPlug() )
+	{
+		ComputeNode::hash( output, context, h );
+		matchPlug()->hash( h );
+	}
+	else if( output == descendantMatchPlug() )
+	{
+		ComputeNode::hash( output, context, h );
+		matchPlug()->hash( h );
+	}
+	else if( output == ancestorMatchPlug() )
+	{
+		ComputeNode::hash( output, context, h );
+		matchPlug()->hash( h );
+	}
+	else if( output == closestAncestorPlug() )
+	{
+		const string location = locationPlug()->getValue();
+		if( !location.empty() )
+		{
+			ScenePlug::PathScope scope( context, ScenePlug::stringToPath( location ) );
+			h = closestAncestorInternalPlug()->hash();
+		}
+		else
+		{
+			h = output->defaultHash();
+		}
+	}
+	else if( output == closestAncestorInternalPlug() )
+	{
+		if( scenePlug()->existsPlug()->getValue() )
+		{
+			const int m = filterPlug()->match( scenePlug() );
+			if( m & PathMatcher::ExactMatch )
+			{
+				ComputeNode::hash( output, context, h );
+				const ScenePlug::ScenePath path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
+				h.append( path.data(), path.size() );
+				return;
+			}
+			else if( m & PathMatcher::AncestorMatch )
+			{
+				ScenePlug::ScenePath path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
+				if( path.size() )
+				{
+					path.pop_back();
+					ScenePlug::PathScope scope( context, path );
+					h = closestAncestorInternalPlug()->hash();
+					return;
+				}
+			}
+		}
+		h = output->defaultHash();
+	}
+}
+
+void FilterQuery::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
+{
+	if( output == matchPlug() )
+	{
+		unsigned match = IECore::PathMatcher::NoMatch;
+		const string location = locationPlug()->getValue();
+		if( !location.empty() )
+		{
+			ScenePlug::PathScope scope( context, ScenePlug::stringToPath( location ) );
+			if( scenePlug()->existsPlug()->getValue() )
+			{
+				match = filterPlug()->match( scenePlug() );
+			}
+		}
+		static_cast<IntPlug *>( output )->setValue( match );
+	}
+	else if( output == exactMatchPlug() )
+	{
+		static_cast<BoolPlug *>( output )->setValue(
+			matchPlug()->getValue() & IECore::PathMatcher::ExactMatch
+		);
+	}
+	else if( output == descendantMatchPlug() )
```

Comment:
Should these plugs be uncached currently, since they're currently just direct wrappers on matchPlug?

Reply:
Added in aeab136.